rfcs/rfc-2-font-measuring.md

```diff
@@ -10,49 +10,54 @@ I propose two different solutions for the issues above.
 1. KayakContext keeps track of all fonts that are loaded. This allows us to use the available measuring to properly size the Text widget.
 2. bevy_kayak_ui uses include_bytes to load in a default font(probably Roboto).
 
-For the first solution an example of how that would look is something like:
+
+## How this will look
+
+We'll need a few new types to store the fonts. I would also like to make something more generic and reusable in the future. It'll look something like this:
+
+`AssetHandle`:
 ```rust
-fonts: HashMap<u16, KayakFont>,
+pub struct AssetHandle<T> {
+    id: u32, // Optionally UUID here although I don't think its required..
+    phantom_data: PhantomData<T>,
+}
 ```
 
-For the second issue an example would look like:
+`AssetStorage`:
 ```rust
-let image = Image::new(
-    size,
-    dimension: TextureDimension::D2,
-    data: include_bytes!("./assets/roboto.png"),
-    format: TextureFormat::Rgba8Unorm,
-);
-let atlas_handle = images_assets.add(image);
-
-let mut font = KayakFont::new(
-    Sdf::from_bytes(include_bytes!("./assets/roboto.kayak_font")),
-    atlas_handle,
-);
+pub struct AssetStorage<T> {
+    assets: HashMap<AssetHandle<T>, T>,
+    // Stores bindings that notify us when changes are made.
+    // Optionally we can make Binding Hash here. Not sure if I like that though..
+    binding: HashMap<AssetHandle<T>, Binding<AssetHandle<T>>>,
+}
 ```
 
-## Additional issues:
-The text widget will need to be re-rendered when it's font is loaded in. I suggest quite a simple way of doing this:
+`KayakContext`:
 ```rust
-fonts: Binding<HashMap<u16, KayakFont>>,
+pub struct KayakContext {
+    ..
+    // Stores AssetStorage generically.
+    assets: Resources,
+}
+
+impl KayakContext {
+    pub fn get_asset<T>(&self, asset_handle: AssetHandle<T>) -> &T {
+        // Throw error if AssetStorage doesn't exist in the asset resources.
+        ..
+    }
+
+    // We want binding here because we need to allow widgets to track changes.
+    pub fn set_asset<T>(&mut self, asset: T) -> Binding<AssetHandle<T>> {
+        ..
+    }
```

Comment:
Is this called by the widget? How does it get access to `T`? Or is this called in some other context?

Reply:
`set_asset` would be called by `bevy_kayak_ui`.

Reply:
Gotcha. So we return the `Binding` just in case the setter needs it? And the widget uses `get_asset` to get their own binding?